### ISSUE TITLE
Allow containers to be added to a running pod

### ISSUE THREAD

**Issue by verb** [kubernetes/kubernetes#37838]
**Is this a request for help?**: no

**What keywords did you search in Kubernetes issues before filing this one?**: "add container"

---

**Is this a BUG REPORT or FEATURE REQUEST?** (choose one): FEATURE REQUEST

While exploring solutions to allow troubleshooting of running pods (#27140) it's becoming clear that the most straightforward solution for adding a container to a running pod will to allow containers to be added to the pod spec. This is sufficiently complex that I'm opening a separate issue to track it.

Requirements for adding a container include:

*   fetch container image when pod is added
*   respect admission restrictions
*   allow introspection of pod state (no hidden containers)

Example:

```
% cat nginx.yaml
apiVersion: v1
kind: Pod
metadata:
  name: nginx
spec:
  containers:
  - image: nginx
    name: nginx
% kubectl create -f nginx.yaml
pod "nginx" created
% kubectl get po 
NAME      READY     STATUS    RESTARTS   AGE
nginx     1/1       Running   0          19s
% kubectl patch pod nginx -p '{"spec":{"containers":[{"name": "debug", "image": "debian", "std
in": true, "tty": true}]}}'                                                                                       
"nginx" patched
% kubectl get po 
NAME      READY     STATUS              RESTARTS   AGE
nginx     1/2       ContainerCreating   0          24s
% kubectl get po 
NAME      READY     STATUS    RESTARTS   AGE
nginx     2/2       Running   0          25s
% kubectl attach -it -c debug nginx 
If you don't see a command prompt, try pressing enter.
root@nginx:/# busybox netstat -ntl
Active Internet connections (only servers)
Proto Recv-Q Send-Q Local Address           Foreign Address         State       
tcp        0      0 0.0.0.0:80              0.0.0.0:*               LISTEN   
```

Problems to solve:

* side effects for pod's resources and QoS class
* lifecycle changes (e.g. to pod status) from new container exiting

**Comment by verb** [comment]
I'll put together a proposal for this.

**Comment by ncdc** [comment]
@smarterclayton re `oc debug` - are we considering upstreaming that from OpenShift?

**Comment by smarterclayton** [comment]
It could be done - it depends a little bit on niceties from openshift like rsh but in general solve this problem pretty well for getting a reproducible environment.

In practice, only a few problems have been insoluble with `oc debug` and were generally related to volume mounts and not remembering to stop other pods.

**Comment by smarterclayton** [comment]
oc debug doesn't help if you don't have a shell in your image, but could easily add that.

**Comment by verb** [comment]
The problem I want to solve in #27140 is debugging an application in situ, but it sounds like `oc debug` makes a copy of a pod. At the risk of generalizing, it sounds like `oc debug` is excellent for troubleshooting pod config but not application level troubleshooting.

There's clearly overlap here, though, and I wonder if we couldn't support both use cases.

**Comment by verb** [comment]
Playing around with it a bit, it looks like adding a container could be some help in troubleshooting pod configuration/startup issues by adding a container to a pod in CrashLoopBackOff pod, for example:

```
% cat crashloop.yaml                                                                      
apiVersion: v1
kind: Pod
metadata:
  name: crashloop
spec:
  containers:
  - name: crash
    image: debian
    args:
    - "false"
% kubectl create -f crashloop.yaml
pod "crashloop" created
% kubectl exec -it crashloop -c crash bash
error: Internal error occurred: error executing command in container: container not found ("crash")
% kubectl patch pod crashloop -p '{"spec":{"containers":[{"name": "debug", "image": "debian", "stdin": true, "tty": true}]}}'
"crashloop" patched
% kubectl exec -it crashloop -c debug bash
root@crashloop:/#
```

Obviously we'd want a more user-friendly experience that `kubectl patch` (covered in #27140) and this is only useful if we do something useful with volumes-from/cross-container mounts (#831).

**Comment by smarterclayton** [comment]
I don't want to get people's hopes up - letting containers be mutable is a pretty big decision and it's unlikely to happen anytime in the next few releases.  If this is important, it's probably better to work with the existing API rather than proposing changes that are almost certainly going to lack review bandwidth to make them happen (review bandwidth is precious for existing problems).

**Comment by loewenstein** [comment]
What is the state of mutable list of containers? Anything outside of this issue I should track?

I am writing a [BOSH CPI for Kubernetes](https://github.com/sap/bosh-kubernetes-cpi-release) and being able to add containers to a running Pod would be a huge step forward (as would be adding volumes to a running Pod).

For those curious, [BOSH](https://bosh.io) is the tool to install [Cloud Foundry](https://www.cloudfoundry.org/) on whatever cloud BOSH has a Cloud Provider Interface (CPI) for. Hence my work is adding Kubernetes to the already huge list of installation targets.

**Comment by verb** [comment]
@loewenstein It's very unlikely this will take the form of a fully mutable list of containers. For the latest design please see kubernetes/community#649. Would something like this meet the requirements for your use case?

**Comment by loewenstein** [comment]
@verb Sorry I didn't find time to respond earlier. The work you are doing looks interesting and I agree that it suffices for the immediate use case of debugging. 

However, it does not help in my case. I really need to change _the real containers_. In a certain sense I am trying to integrate a "vm scheduler" with kube. I.e. BOSH has a similar feature set to Kubernetes. The users defines the desired state of a distributed  software system and BOSH takes care that the actual state matches. It requests VMs and disks from IaaSes, installs a base OS, software on top and starts processes. It continuously monitors that everything is fine and if necessary intervenes and fixes any issues. Unfortunately it has separate APIs for creating VMs and starting processes on them. 

Is there huge opposition against having mutable Pods? Or did you choose your particular implementation path because it is really easier to implement? If there is opposition, could you tell me where I should start lobbying?

**Comment by verb** [comment]
After discussing the goals of SIG Node, the clear consensus is that the containers list in the pod spec should remain immutable. #27140 will be better addressed by kubernetes/community#649, which allows running an ephemeral debugging container in an existing pod. This will not be implemented.
